src/control/server/ctl_storage_rpc.go

```diff
@@ -769,11 +769,16 @@ func (cs *ControlService) StorageScan(ctx context.Context, req *ctlpb.StorageSca
 	return resp, nil
 }
 
-func (cs *ControlService) formatMetadata(instances []Engine, reformat bool) (bool, error) {
+func (cs *ControlService) formatMetadata(instances []Engine, reformat, replace bool) (bool, error) {
 	// Format control metadata first, if needed
 	if needs, err := cs.storage.ControlMetadataNeedsFormat(); err != nil {
 		return false, errors.Wrap(err, "detecting if metadata format is needed")
 	} else if needs || reformat {
+		if replace == true {
+			return false, errors.New("control metadata formatting is not possible " +
+				"when rank replace is requested")
+		}
```

Comment:
Hmm, I guess I assumed you might do something special to support this case, like repair the affected engine's metadata subdirectory, instead of rejecting it outright. Since this change is to address a pmem problem, I'm fine with not supporting MD-on-SSD right away. Might be better to check if it's an MD-on-SSD config immediately after getting the format request, and deny any `--replace` request in that case.

Reply:
I haven't worked out the details of the MD-on-SSD case yet as I haven't seen it in the wild, thinking about it if we have a media failure on a control_metadata storage device the whole thing would need to be rewritten anyway so not sure it's necessary to have the ability to rewrite individual engine control_metadata for this case. Maybe we could allow control_metadata format only in the case where both ranks require format and replace is requested. This would involve relaxing the restriction on replace not being allowed if both ranks require format (which is one of your other review comments) and then rely on the `joinReplace()` logic to reject inappropriate replacement requests.

Reply:
That makes sense to me.